src-docs/docs/components/server/config.md
![META](authors:Nicolas "Duduf" Dufresne;license:GNU-FDL;copyright:2021-2022;updated:2022/11/01)

# Server configuration

The server can and must be configured by editing these two files: `config/config.php` and `config/config_logs.php`.

## config.php

The `config.php` file contains the main configuration and should look like that:

```php
<?php
    /*
		Ramses: Rx Asset Management System
        
        This program is licensed under the GNU General Public License.

        Copyright (C) 2020-2023 Nicolas Dufresne and Contributors.

        This program is free software;
        you can redistribute it and/or modify it
        under the terms of the GNU General Public License
        as published by the Free Software Foundation;
        either version 3 of the License, or (at your option) any later version.

        This program is distributed in the hope that it will be useful,
        but WITHOUT ANY WARRANTY; without even the implied warranty of
        MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
        See the GNU General Public License for more details.

        You should have received a copy of the *GNU General Public License* along with this program.
        If not, see http://www.gnu.org/licenses/.
	*/

	// Edit this configuration file before running the install script at /install/index.php

	// === DEV MODE & DEBUG Info ===

	// Activates printing the SQL & PHP errors.
	// For security reasons, it is important to set this to false in production mode
	$devMode = false;
	// Sets the minimum level of the logs in the server replies sent to the clients
	// One of: 'DATA', 'DEBUG', 'INFO', 'WARNING', 'CRITICAL', 'FATAL'
	$logLevel = 'WARNING';

	// ==== SQL SETTINGS ====

	// either 'mysql', 'mariadb' or 'sqlite'
	// Note: with 'mysql', the minimum version of MySQL is 8.0.19
	$sqlMode = 'sqlite';
	// Table prefix
	// should be a random character string (use https://duckduckgo.com/?q=generate+password)
	// you can setup multiple instances on the same DB
	// if each use a different prefix
	$tablePrefix = "5H3VBzSY";

	// ==== MySQL/MariaDB SETTINGS ====

	// Host URL
	$sqlHost = "db";
	$sqlPort = 3306;
	// Database name
	$sqlDBName = "ramses";
	// User
	$sqlUser = "ramses";
	// Password
	$sqlpassword = "password";

	// ==== SESSION & SECURITY SETTINGS ====

	// This must be the server public adress, exactly as used in the clients
	$serverAddress = "localhost/ramses";

	// Whether to accept only SSL connections
	// This should always be true, except maybe on dev environments.
	$forceSSL = true;

	// Compression may cause issues on some servers,
	// Set this to true to deactivate it.
	$disableCompression = false;

	// Session timeout (seconds)
	// The client will be disconnected after being idle for this time
	// 30 minutes by default (1800) are more than enough:
	// The official client makes a call at least every 2 minutes by default.
	// A negative value completely deactivates the timeout check.
	$sessionTimeout = 1800;
	// Max Session timeout (seconds)
	// The client will be disconnected no matter what after this time
	// 12 hours by default( 43200 )
	$cookieTimeout = 43200;

	// This should never be changed, unless you change the key before building the official client or implementing your own client.
	// It can be used to make sure only your own client, built by yourself, can connect to your own server. In this case, keep it secret!
	// It is used to hash passwords.
	$clientKey = "drHSV2XQ";
?>
```

- `$devMode` should always be `false` unless this server is going to be used for development and tests in a development environement. For security reasons **do not set this to `true`** on a production server.
- `$logLevel` controls the quantity of logs included in the server replies and displayed by *Ramses Applications*. This is independant from the logs saved on the server configured by the `config/config_logs.php` file.
- `$sqlMode` can be either `'sqlite'` or `'mysql'` or `'mariadb'` depending on your choice of configuration (read above).
- `$tablePrefix` is a prefix which will be used for the name of the database tables used by *Ramses*. It allows you to share a database with other instances of *Ramses* and even any other application (for example if your provider doesn't allow you to create more databases on your server).

These next options are only used in *MySQL* mode. Ignore them when using *SQLite*.

- `$sqlHost` is the host name or IP of your *MySQL* server. Don't forget the quotes!
- `$sqlPort` is the port used by the *MySQL* server. `3306` is the default for most installations, if you need something different, your hosting provider should tell you.
- `$sqlDBName` is the name of the specific database you have to create for *Ramses*. 
- `$sqlUser` is the name of the specific *MySQL* user you have to create for *Ramses*.
- `$sqlpassword` is the password of the specific *MySQL* user you have to create for *Ramses*.

!!! note
	When using the *MySQL Docker* folder, do not change the `sqlHost` value, unless you've changed the settings in the `docker_compose.yml` file.

And finally complete these settings too:

- `$serverAddress` must be the one and only address used by the *Ramses Applications* to connect to this server. If this is incorrectly set, *Ramses Applications* won't be able to connect to this server.
- `$forceSSL` will always redirect calls to `https` when set to `true`. Set this to `false` if you don't have any SSL certificate on this server (not recommended).
- `$disableCompression` allows to disable the compression of the data when it is sent to the client (not recommended). Change this only if there are issues with the connexion to the client or if you know your server can't send compressed data (using gzip).
- `$sessionTimeout` is the time, in seconds, before an idle client is disconnected.
- `$cookieTimeout` is the time, in seconds, before a client is disconnected, no matter what.
- `$clientKey` can be changed *only* if your client is not the official open source *Ramses Client Application*. Otherwise, leave it to the default value. It is a way to make sure only your client can connect to the server. It is currently used only when setting up the first user (administrator) during the server installation.

## config_logs.php

*Ramses* can log what happens on the server. This can be useful for debugging or to diagnose issues on the server. By default, all logs are deactivated.

Logs are stored by date in the `logs` subfolder inside the ramses server folder. Connection logs and debug logs are stored as *CSV* files (one per day), request logs are *text* and *json* files (one subfolder per request).

This is the configuration for the logs:

```php
<?php
    /*
		Ramses: Rx Asset Management System
        
        This program is licensed under the GNU General Public License.

        Copyright (C) 20202-2021 Nicolas Dufresne and Contributors.

        This program is free software;
        you can redistribute it and/or modify it
        under the terms of the GNU General Public License
        as published by the Free Software Foundation;
        either version 3 of the License, or (at your option) any later version.

        This program is distributed in the hope that it will be useful,
        but WITHOUT ANY WARRANTY; without even the implied warranty of
        MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
        See the GNU General Public License for more details.

        You should have received a copy of the *GNU General Public License* along with this program.
        If not, see http://www.gnu.org/licenses/.
	*/

	// Edit this configuration file to adjust server logging

	// === GLOBALLY ENABLE LOGS ===
	$enableLogs = false;

    // Connexion logs
    $connexionLogs = false;
    // Requests logs
    $requestLogs = false;
    // Debug logs
    $debugLogs = false;

    // How long do we keep logs? (days)
    $logsExpiration = 2;
?>
```

- `$enableLogs`: set this to `true` or `false` to globally enable logs.
- `$connexionLogs`: when `true`, ramses will log all connections and disconnections to and from the server.
- `$requestLogs`: when `true`, ramses will log requests received by the server including: the request headers, body, and some metadata. The details of the request headers may change according to the web server you're using and its version.
- `$debugLogs`: when `true`, ramses will log some extra information coming with each request.
- `$logsExpiration` sets how many days the logs will be kept. Keep this value small if there's a lot of activity on your server to avoid quickly growing the log size.

src-docs/docs/components/client/settings.md
![META](authors:Nicolas "Duduf" Dufresne;license:GNU-FDL;copyright:2021-2022;updated:2022/02/07)

# ![](/img/icons/gotosettings.svg){: style="width:55px;"} Application settings and overview

<figure>
  <img src="/img/client/home.png"/>
  <figcaption>The home screen, shown when starting the application.</figcaption>
</figure>

When running the application for the first time, you need to go to the settings first to configure it.

## Overview

### General shortcuts and behaviour

In all the application, a few shortcuts are always the same, in addition to your system's usual shortcuts for *copy*, *paste*, *undo* etc.

- `[Middle Click]`: scroll/pan in lists, tables, timeline...
- `[Ctrl] + [Middle Click]` or `[Mouse Wheel]`: zoom in or out in views which can be zoomed.
- `[Shift] + [A]`: add/create something.
- `[Del]` or `[Shift] + [X]`: remove/delete selection.

All side panels (docks) can be moved around and docked to the sides or undocked by clicking on their titles.

Menu bars can also be moved to all sides of the application.

### Title bar

![](/img/client/titlebar.png)

The title bar show the main buttons to switch the panels on the left, and the visibility buttons on the right. Click and drag an empty space of the title bar to move the whole window.

![](/img/client/projectselector.png)

In the middle of the title bar is the selector where you can select the project you're currently working on.

![](/img/icons/hide_sd.svg){: style="width:18px;"} Hide. The application will not close but sit in the system tray.

Which panels and buttons are available depends on [the **role** of the current user](../../pipeline/administration.md), and there may be fewer buttons than on the picture above.

- ![](/img/icons/admin.svg){: style="width:18px;"} [Administration](admin.md) panel (*Administrators* only).
- ![](/img/icons/cinema-movie-settings_bl.svg){: style="width:18px;"} [Project settings](project.md) panel (*Project administrators* and *Administrators* only).
- ![](/img/icons/edit-connections-pipeline_bl.svg){: style="width:18px;"} [Pipeline editor](pipeline.md) panel (*Project administrators* and *Administrators* only).
- ![](/img/icons/assets_sl.svg){: style="width:18px;"} [Assets table](production.md) panel.
- ![](/img/icons/shots.svg){: style="width:18px;"} [Shots table](production.md) panel.
- ![](/img/icons/calendar_sl.svg){: style="width:18px;"} [Schedule](schedule.md) panel.
- ![](/img/icons/counter_bl.svg){: style="width:18px;"} Statistics panel.

### Status bar

![](/img/client/statusbar.png)

The status bar at the bottom of the window displays temporary feedback, useful messages and help on its left section. On the right section, there are a few secondary buttons.

- ![](/img/icons/reload_sl.svg){: style="width:18px;"} The *reload* button updates all data from the server. You should not need to click this button too often as the data is automatically and regularly pulled from the server.
- ![](/img/icons/bash-terminal_sl.svg){: style="width:18px;"} The *console* button shows the console panel (see the next section) where *Ramses* logs useful information for debugging or for developpers. Errors and warnings are also logged there.
- ![](/img/icons/warning_bl.svg){: style="width:18px;"} | ![](/img/icons/critical_bl.svg){: style="width:18px;"} A warning or an error button may be shown next, when there are some issues. Click on the button to show the latest logs and learn what's wrong; a *clear* button is available on the popup to remove the warning.
- The ***online***/***offline*** button can be used to switch between online and offline mode. Going online may need to input your password again.
- The ***user*** button (which shows your user name, or *Guest* if you're not logged in yet) gives access to a few user options (see below).
- ![](/img/icons/gotosettings.svg){: style="width:18px;"} The *settings* button leads you to the settings of the application.
- ![](/img/icons/help.svg){: style="width:18px;"} The *help* button shows a small menu with useful shortcuts.

### Console

![](/img/client/console.png)

The console panel is where *Ramses* logs useful information for debugging or for developpers. Errors and warnings are also logged there, respectively in red and yellow.

As many other *Ramses* panel, the console is a *docked* panel which can be undocked or docked both on the right and left side of *Ramses* by just clicking its title and moving it.

![](/img/client/undock.png)

### User Profile

Use the *user menu* at the bottom right of the window (the button with your ID) to access your profile settings.

![](/img/client/profile.png)

Except for the password, there's no need to validate any change; changes are saved instantly.

- ***Name*** is your display name and can be changed any time (although it's not really advisable).
- ***ID*** is your unique identifier which is used in the files and folders (see [*Ramses Tree*](../files/index.md)). It can't be changed once your user folder has been created. Read the [*Ramses naming conventions*](../files/naming.md) to learn how *IDs* and names have to be formatted.
- ***Comment*** can be any text associated to your user. It is public.
- The ***Current role*** can never be changed by the user itself, and has to be changed by an administrator.
- To update your password, input your ***Current password*** and repeat your ***New password*** twice. Don't forget to click the ***Change password*** button to validate!
- You can change the ***folder*** where your personnal files are stored.
    - Leave the field empty to set back the default folder.
    - The blue ![](/img/icons/set-folder_sl.svg){: style="width:18px;"} button opens a folder selector to change the folder location.
    - The green ![](/img/icons/go-to-explore-reveal-folder_sl.svg){: style="width:18px;"} button opens the folder in your file explorer.

### Lists

![](../../img/client/list.png)

There are a lot of lists in *Ramses*...

In all lists, you can grab the bottom margin of any row to reveal more details. Clicking on a item opens the corresponding properties in a side panel. If you need to select several items, you can click and hold while dragging the mouse to select contiguous items, or you can hold `[Ctrl]` while clicking items to select individual items.

The search bar uses both IDs and names to find items.

In most of the lists, `[Del]` and `[Shift] + [X]` can be used to remove items. `[Middle Click]` can be used to scroll the list, as the `[Mouse Wheel]` does too.

### Comments

![](../../img/client/comment-formatted.png)

In many places, you can add *Comments* in *Ramses*. These comments can be simple texts, but they can also be formatted using the [*Markdown*](https://www.markdownguide.org/) syntax.

Start by writing some text, and validate your input with `[Enter]` from the numpad. *Ramses* will interpret the text to display a nice rich text.

You can get back to the source to edit the text in more details using the `[F2]` key.

![](../../img/client/comment-source.png)

Use the `[Right Click]` context menu for other options.

![](../../img/client/comment-format-context-menu.png)

## ![](/img/icons/gotosettings.svg){: style="width:48px;"} Settings

To open the settings panel, click on the ![](/img/icons/gotosettings.svg){: style="width:18px;"} settings button at the bottom right corner of the window.

!!! Note
    These settings are local settings, and will not be shared with other workstations. Each workstation, (actually, each session on the system) has its own settings.

The purple ![](/img/icons/reinit.svg){: style="width:18px;"} button will restore all default settings.

Use the blue ![](/img/icons/close.svg){: style="width:18px;"} close button to close the settings. They are automatically saved.

### ![](/img/icons/color.svg){: style="width:32px;"} Appearance

![](/img/client/appearancesettings.png)

The first tab shows the appearance settings; you can change these as you wish.

### ![](/img/icons/storage_sl.svg){: style="width:32px;"} Ramses Data

![](/img/client/datasettings.png)

The second tab defines how *Ramses* will store the data on the filesystem, and has a few display option for the data.

- ***Ramses main folder*** defines the location of the main *Ramses* root folder, the one which will contain the `Config`, `Projects` and `Users` subfolders, as defined by the [*Ramses Tree*](../files/index.md).  
    By default, this folder is located in your user folder.

The blue ![](/img/icons/set-folder_sl.svg){: style="width:18px;"} button opens a folder selector to change the folder location.

The green ![](/img/icons/go-to-explore-reveal-folder_sl.svg){: style="width:18px;"} button opens the folder in your file explorer.

!!! Note
    Project folders and user folders can be individually changed later to be outside of this main *Ramses* root folder, for example if you need to locate specific projects in other arbitrary specific folders. But *Ramses* still needs this main folder to be defined, at least to store some configuration files.

- ***Date format*** defines how dates will be displayed in the user interface. It does not change anything about how the data is actually stored in the database and files.

- ***Time format*** defines how times will be displayed in the user interface. It does not change anything about how the data is actually stored in the database and files.

### ![](/img/icons/server-settings.svg){: style="width:32px;"} Server

![](/img/client/serversettings.png)

The third tab configures how the application will connect to the server.

- ***Server address*** must be the complete *URL* to the server.

- ***Secure connexion*** must be checked in case you want to encrypt the data sent to and recieved from the server. Your server must be compatible with *SSL* and have a valid certificate.

- ***Update every*** *X minutes* defines how often the application will automatically update all its data from the server. In order to keep the application very lightweight and the server very light and easy to install on all kind of computers, the local data is not updated in realtime (there's no pushing from the server), but the application will regularly pull changes from the server. This value does not need to be too low, as *Ramses* does not really need to be perfectly up-to-date every second. __The default value is *2 minutes*__.

- ***Server timeout*** defines how long the application will wait for the server response when getting or posting data. If this timeout expires, the server will be considered not working correctly and the application will switch to offline mode. You may raise a bit this value if your server or your connexion are particularly slow. __The default value is *3 seconds*__.

### ![](/img/icons/daemon.svg){: style="width:32px;"} Daemon

![](/img/client/daemon.png)

The *Daemon* is a background service, a small local server, which is started with the application and is in charge of communications with the [*Ramses Add-ons*](../addons/index.md).

- ***Listenning on port*** defines the local port used by the *Daemon* to receive connections and requests from the [*Ramses Add-ons*](../addons/index.md) and other services. You may change this value if it conflicts with other applications; in this case, you'll have to change it in the settings of all the [*Ramses Add-ons*](../addons/index.md) too. **The default value is _18185_**[^1].

- Click the ***Restart Daemon*** button when changing the listenning port to apply the change.

[^1]:
    This value has been chosen after the birth date of [Karl Marx](https://en.wikipedia.org/wiki/Karl_Marx), 5 May 1818.

Ramses-Server/src/config/config.php
<?php
    /*
		Ramses: Rx Asset Management System
        
        This program is licensed under the GNU General Public License.

        Copyright (C) 20202-2021 Nicolas Dufresne and Contributors.

        This program is free software;
        you can redistribute it and/or modify it
        under the terms of the GNU General Public License
        as published by the Free Software Foundation;
        either version 3 of the License, or (at your option) any later version.

        This program is distributed in the hope that it will be useful,
        but WITHOUT ANY WARRANTY; without even the implied warranty of
        MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
        See the GNU General Public License for more details.

        You should have received a copy of the *GNU General Public License* along with this program.
        If not, see http://www.gnu.org/licenses/.
	*/

	// Edit this configuration file before running the install script at /install/index.php

	// === DEV MODE ===

	// Activates printing the SQL & PHP errors.
	// For security reasons, it is important to set this to false in production mode
	$devMode = false;

	// ==== SQL SETTINGS ====

	// either 'mysql' (works for MariaDB too) or 'sqlite'
	$sqlMode = 'sqlite';

	// ==== MySQL/MariaDB SETTINGS ====

	// Host URL
	$sqlHost = "localhost";
	$sqlPort = 3306;
	// Database name
	$sqlDBName = "ramses";
	// User
	$sqlUser = "ramses";
	// Password
	$sqlpassword = "password";
	// Table prefix
	// should be a random character string (use https://duckduckgo.com/?q=generate+password)
	// you can setup multiple instances on the same DB
	// if each use a different prefix
	$tablePrefix = "5H3VBzSY";

	// ==== SESSION & SECURITY SETTINGS ====

	// Whether to accept only SSL connections
	// This should always be true, except maybe on dev environments.
	$forceSSL = true;

	// Session timeout (seconds)
	// The client will be disconnected after being idle for this time
	// 30 minutes by default (1800) are more than enough:
	// The official client makes a call at least every 2 minutes by default.
	$sessionTimeout = 1800;
	// Max Session timeout (seconds)
	// The client will be disconnected no matter what after this time
	// 12 hours by default( 43200 )
	$cookieTimeout = 43200;

	// This must be the server public adress, exactly as used in the clients
	$serverAddress = "localhost/ramses";

	// This should never be changed, unless you change the key before building the official client or implementing your own client.
	// It is used to break compatibility between some updates of the client/server framework
	// It can be used to make sure only your own client, built by yourself, can connect to your own server. In this case, keep it secret!
	// It is used to hash passwords.
	$clientKey = "drHSV2XQ";
?>
